### ISSUE TITLE
IP format checking not correct in local dns client

### ISSUE THREAD

**Issue by kyze8439690** [XTLS/Xray-core#2617]
Problem line: https://github.com/XTLS/Xray-core/blob/e244db76fbf871cc61f2f4d99f2676be27c2d9e8/features/dns/localdns/client.go#L36

IPv4 format address is recognized as ipv6 format address.

Code below should fix this issue:

```go
for _, ip := range ips {
	parsed := net.IPAddress(ip)
	if parsed != nil {
		parsedIPs = append(parsedIPs, parsed.IP())
	}
	if ip.To4() != nil {
		ipv4 = append(ipv4, ip)
	}
	if ip.To16() != nil {
		ipv6 = append(ipv6, ip)
	}
}
```

**Comment by chika0801** [comment]
Thanks for the fix, you can send a request at https://github.com/XTLS/Xray-core/pulls.

**Comment by kyze8439690** [comment]
> I don't think your fix will work because `ip.To16()` will always success for IPv4 address.

But it always  check `ip.To4()` first.

**Comment by us254** [comment]
```

for _, ip := range ips {
	parsed := net.ParseIP(ip)
	if parsed != nil {
		switch {
		case parsed.To4() != nil:
			ipv4 = append(ipv4, parsed)
		case parsed.To16() != nil:
			ipv6 = append(ipv6, parsed)
		}
	}
}

```

**Comment by kyze8439690** [comment]
> Although `len(ip)` is not a good check, can you point out how to trigger `net.Lookup` to return non-4-byte IPv4 addresses for a domain in this use case? golang/go@c53390b

Android 13, trigger dns query on a ipv4 only network.

**Comment by kyze8439690** [comment]
> Although `len(ip)` is not a good check, can you point out how to trigger `net.Lookup` to return non-4-byte IPv4 addresses for a domain in this use case? [golang/go@c53390b](https://github.com/golang/go/commit/c53390b)

I think this commit do fix this issue, after I update golang to 1.21.2 and rebuild the core, this issue is gone.
